File: scripts/scenarios/scenarios/02_social_graph.py
"""Scenario 2: "Making Friends" — Social Graph

Marcus follows Luna, Rosa, and DJ Volt. Luna follows Marcus back.
Quiet Observer follows everyone. Rosa follows Luna and Marcus.
DJ Volt follows Rosa. Marcus unfollows DJ Volt. Luna blocks Trollface.

Services: PDS, AppView (optional — some endpoints skip if no AppView)
"""

from __future__ import annotations

import sys
import time
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import XrpcClient, get_character, PDS1, ScenarioResult, timed_call



def _now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def _create_accounts(client: XrpcClient, names: list[str], result: ScenarioResult) -> dict:
    sessions = {}
    for name in names:
        char = get_character(name)
        session = timed_call(
            result, f"Create account: {char.name}",
            lambda c=char: client.accounts.create_account(c.handle, c.email, c.password),
            detail_fn=lambda s, n=name: f"did={s['did']}",
        )
        if session:
            char.did = session["did"]
            char.access_jwt = session["accessJwt"]
            sessions[name] = session
    return sessions


def _follow(client: XrpcClient, follower_name: str, target_name: str, result: ScenarioResult) -> dict | None:
    follower = get_character(follower_name)
    target = get_character(target_name)

    if not follower.did or not follower.access_jwt:
        result.step_skipped(f"{follower.name} follows {target.name}", "Follower account not created")
        return None
    if not target.did:
        result.step_skipped(f"{follower.name} follows {target.name}", "Target account not created")
        return None

    rec = timed_call(
        result, f"{follower.name} follows {target.name}",
        lambda: client.records.create_record(
            follower.did, "app.bsky.graph.follow",
            {"$type": "app.bsky.graph.follow", "subject": target.did, "createdAt": _now()},
            follower.access_jwt),
        detail_fn=lambda r: f"uri={r['uri']}",
    )
    return rec


def run() -> ScenarioResult:
    result = ScenarioResult("Social Graph")
    result.start()

    client = XrpcClient(PDS1)

    timed_call(result, "Server health check",
               lambda: client.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    pds1_chars = ["luna", "marcus", "rosa", "volt", "troll", "quiet", "admin"]
    _create_accounts(client, pds1_chars, result)

    active = [n for n in pds1_chars if get_character(n).did]
    if len(active) < 4:
        result.step_failed("Account creation", f"Only {len(active)} accounts created, need at least 4")
        result.finish()
        return result

    for name in active:
        char = get_character(name)
        timed_call(
            result, f"Set profile: {char.name}",
            lambda c=char: client.records.create_record(
                c.did, "app.bsky.actor.profile",
                {"$type": "app.bsky.actor.profile", "displayName": c.name, "description": c.persona},
                c.access_jwt),
            skip_on_status={404},
        )

    _follow(client, "marcus", "luna", result)
    _follow(client, "marcus", "rosa", result)
    _follow(client, "marcus", "volt", result)
    _follow(client, "luna", "marcus", result)
    for name in ["luna", "marcus", "rosa", "volt", "troll", "admin"]:
        _follow(client, "quiet", name, result)
    _follow(client, "rosa", "luna", result)
    _follow(client, "rosa", "marcus", result)
    _follow(client, "volt", "rosa", result)

    time.sleep(2)

    marcus = get_character("marcus")
    luna = get_character("luna")

    timed_call(
        result, "Marcus's follows list",
        lambda: client.graph.get_follows(marcus.did, token=marcus.access_jwt),
        detail_fn=lambda f: f"count={len(f.get('follows', []))}",
    )

    timed_call(
        result, "Luna's followers list",
        lambda: client.graph.get_followers(luna.did, token=luna.access_jwt),
        detail_fn=lambda f: f"count={len(f.get('followers', []))}",
    )

    volt = get_character("volt")
    follows_resp = timed_call(
        result, "Marcus lists follow records (for unfollow)",
        lambda: client.records.list_records(marcus.did, "app.bsky.graph.follow", token=marcus.access_jwt),
    )
    if follows_resp:
        records = follows_resp.get("records", [])
        volt_follow = None
        for rec in records:
            val = rec.get("value", {})
            if val.get("subject") == volt.did:
                volt_follow = rec
                break
        if volt_follow:
            rkey = volt_follow["uri"].split("/")[-1]
            timed_call(
                result, "Marcus unfollows DJ Volt",
                lambda: client.records.delete_record(marcus.did, "app.bsky.graph.follow", rkey, marcus.access_jwt),
                detail_fn=lambda r: f"deleted rkey={rkey}",
            )
        else:
            result.step_skipped("Marcus unfollows DJ Volt", "Follow record not found")

    troll = get_character("troll")
    if not luna.did or not luna.access_jwt:
        result.step_skipped("Luna blocks Trollface", "Luna account not created")
    elif not troll.did:
        result.step_skipped("Luna blocks Trollface", "Troll account not created")
    else:
        timed_call(
            result, "Luna blocks Trollface",
            lambda: client.records.create_record(
                luna.did, "app.bsky.graph.block",
                {"$type": "app.bsky.graph.block", "subject": troll.did, "createdAt": _now()},
                luna.access_jwt),
            detail_fn=lambda r: f"uri={r['uri']}",
        )

    if not luna.did or not luna.access_jwt:
        result.step_skipped("Luna's blocks list", "Luna account not created")
    else:
        timed_call(
            result, "Luna's blocks list",
            lambda: client.graph.get_blocks(luna.access_jwt),
            detail_fn=lambda b: f"count={len(b.get('blocks', []))}",
        )

    timed_call(
        result, "Marcus profile counts",
        lambda: client.feed.get_profile(marcus.did, token=marcus.access_jwt),
        detail_fn=lambda p: f"follows={p.get('followsCount', 0)}, followers={p.get('followersCount', 0)}",
    )

    if not luna.did or not luna.access_jwt:
        result.step_failed("Search actors", "Luna account not created")
    else:
        timed_call(
            result, "Search actors",
            lambda: client.feed.search_actors("Luna", token=luna.access_jwt),
            detail_fn=lambda s: f"found={len(s.get('actors', []))}",
        )

    result.finish()
    return result


if __name__ == "__main__":
    result = run()
    result.print_summary()
    sys.exit(result.exit_code)

File: scripts/scenarios/scenarios/02_social_graph.ts
/**
 * @module scenarios/02_social_graph
 *
 * Scenario: Social Graph management (follows, blocks, and counts)
 *
 * Behavior:
 * - Create test accounts and profiles
 * - Establish a follower graph between test users
 * - Marcus follows/unfollows DJ Volt
 * - Luna blocks Trollface
 * - Verify follow/follower counts and lists
 *
 * Expectations:
 * - Follow records are created and deleted correctly
 * - Block records are created
 * - Actor lists and counts reflect the social graph accurately
 */

import { XrpcClient } from "@garazyk/gruszka";
import type { ScenarioContext } from "@garazyk/hamownia/config";
import { createScenarioContext } from "@garazyk/hamownia/scenario-context";
import { ScenarioResult, timedCall } from "@garazyk/hamownia";
export { ScenarioResult, StepResult, StepStatus } from "@garazyk/hamownia";
export type { ScenarioReport } from "@garazyk/hamownia";

function now() {
  return new Date().toISOString();
}

async function createAccounts(
  ctx: ScenarioContext,
  client: XrpcClient,
  names: string[],
  result: ScenarioResult,
) {
  const sessions: Record<string, any> = {};
  for (const name of names) {
    const char = ctx.getCharacter(name);
    const session = await timedCall(
      result,
      `Create account: ${char.name}`,
      async () => {
        try {
          const res = await client.agent.createAccount({
            handle: char.handle,
            email: char.email,
            password: char.password,
          });
          return res.data;
        } catch (e: any) {
          if (e.message && e.message.includes("already exists")) {
            const res = await client.agent.login({
              identifier: char.handle,
              password: char.password,
            });
            return res.data;
          }
          throw e;
        }
      },
      (s) => `did=${s.did}`,
    );
    if (session) {
      char.did = session.did;
      char.accessJwt = session.accessJwt;
      sessions[name] = session;
    }
  }
  return sessions;
}

async function follow(
  ctx: ScenarioContext,
  client: XrpcClient,
  followerName: string,
  targetName: string,
  result: ScenarioResult,
) {
  const follower = ctx.getCharacter(followerName);
  const target = ctx.getCharacter(targetName);

  if (!follower.did || !follower.accessJwt) {
    result.stepSkipped(
      `${follower.name} follows ${target.name}`,
      "Follower account not created",
    );
    return null;
  }
  if (!target.did) {
    result.stepSkipped(
      `${follower.name} follows ${target.name}`,
      "Target account not created",
    );
    return null;
  }

  const rec = await timedCall(
    result,
    `${follower.name} follows ${target.name}`,
    async () => {
      const res = await client.raw.post("com.atproto.repo.createRecord", {
        repo: follower.did,
        collection: "app.bsky.graph.follow",
        record: {
          $type: "app.bsky.graph.follow",
          subject: target.did,
          createdAt: now(),
        },
      }, follower.accessJwt);
      return res;
    },
    (r) => `uri=${r.uri}`,
  );
  return rec;
}

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */
export async function run(ctx: ScenarioContext): Promise<ScenarioResult> {
  const result = new ScenarioResult("Social Graph");
  result.start();

  const client = new XrpcClient(ctx.pds1);

  await timedCall(
    result,
    "Server health check",
    async () => {
      const res = await fetch(`${ctx.pds1}/xrpc/com.atproto.server.describeServer`);
      if (!res.ok) throw new Error("Server not healthy");
    },
  );

  if (result.failed > 0) {
    result.finish();
    return result;
  }

  const pds1Chars = [
    "luna",
    "marcus",
    "rosa",
    "volt",
    "troll",
    "quiet",
    "admin",
  ];
  await createAccounts(ctx, client, pds1Chars, result);

  const active = pds1Chars.filter((n) => ctx.getCharacter(n).did);
  if (active.length < 4) {
    result.stepFailed(
      "Account creation",
      `Only ${active.length} accounts created, need at least 4`,
    );
    result.finish();
    return result;
  }

  for (const name of active) {
    const char = ctx.getCharacter(name);
    await timedCall(
      result,
      `Set profile: ${char.name}`,
      async () => {
        const res = await client.raw.post("com.atproto.repo.createRecord", {
          repo: char.did,
          collection: "app.bsky.actor.profile",
          record: {
            $type: "app.bsky.actor.profile",
            displayName: char.name,
            description: char.persona,
          },
        }, char.accessJwt);
        return res;
      },
    );
  }

  await follow(ctx, client, "marcus", "luna", result);
  await follow(ctx, client, "marcus", "rosa", result);
  await follow(ctx, client, "marcus", "volt", result);
  await follow(ctx, client, "luna", "marcus", result);
  for (const name of ["luna", "marcus", "rosa", "volt", "troll", "admin"]) {
    await follow(ctx, client, "quiet", name, result);
  }
  await follow(ctx, client, "rosa", "luna", result);
  await follow(ctx, client, "rosa", "marcus", result);
  await follow(ctx, client, "volt", "rosa", result);

  await new Promise((r) => setTimeout(r, 2000));

  const marcus = ctx.getCharacter("marcus");
  const luna = ctx.getCharacter("luna");

  await timedCall(
    result,
    "Marcus's follows list",
    async () => {
      const res = await client.raw.get(
        "app.bsky.graph.getFollows",
        { actor: marcus.did },
        marcus.accessJwt,
      );
      return res;
    },
    (f) => `count=${f.follows?.length || 0}`,
  );

  await timedCall(
    result,
    "Luna's followers list",
    async () => {
      const res = await client.raw.get(
        "app.bsky.graph.getFollowers",
        { actor: luna.did },
        luna.accessJwt,
      );
      return res;
    },
    (f) => `count=${f.followers?.length || 0}`,
  );

  const volt = ctx.getCharacter("volt");
  const followsResp = await timedCall(
    result,
    "Marcus lists follow records (for unfollow)",
    async () => {
      const res = await client.agent.com.atproto.repo.listRecords({
        repo: marcus.did,
        collection: "app.bsky.graph.follow",
      });
      return res.data;
    },
  );

  if (followsResp) {
    const records = followsResp.records || [];
    let voltFollow = null;
    for (const rec of records) {
      if ((rec.value as any).subject === volt.did) {
        voltFollow = rec;
        break;
      }
    }
    if (voltFollow) {
      const rkey = voltFollow.uri.split("/").pop()!;
      await timedCall(
        result,
        "Marcus unfollows DJ Volt",
        async () => {
          await client.raw.post("com.atproto.repo.deleteRecord", {
            repo: marcus.did,
            collection: "app.bsky.graph.follow",
            rkey,
          }, marcus.accessJwt);
        },
        () => `deleted rkey=${rkey}`,
      );
    } else {
      result.stepSkipped("Marcus unfollows DJ Volt", "Follow record not found");
    }
  }

  const troll = ctx.getCharacter("troll");
  if (!luna.did || !luna.accessJwt) {
    result.stepSkipped("Luna blocks Trollface", "Luna account not created");
  } else if (!troll.did) {
    result.stepSkipped("Luna blocks Trollface", "Troll account not created");
  } else {
    await timedCall(
      result,
      "Luna blocks Trollface",
      async () => {
        const res = await client.raw.post("com.atproto.repo.createRecord", {
          repo: luna.did,
          collection: "app.bsky.graph.block",
          record: {
            $type: "app.bsky.graph.block",
            subject: troll.did,
            createdAt: now(),
          },
        }, luna.accessJwt);
        return res;
      },
      (r) => `uri=${r.uri}`,
    );
  }

  if (!luna.did || !luna.accessJwt) {
    result.stepSkipped("Luna's blocks list", "Luna account not created");
  } else {
    await timedCall(
      result,
      "Luna's blocks list",
      async () => {
        const res = await client.raw.get(
          "app.bsky.graph.getBlocks",
          {},
          luna.accessJwt,
        );
        return res;
      },
      (b) => `count=${b.blocks?.length || 0}`,
    );
  }

  await timedCall(
    result,
    "Marcus profile counts",
    async () => {
      const res = await client.raw.get(
        "app.bsky.actor.getProfile",
        { actor: marcus.did },
        marcus.accessJwt,
      );
      return res;
    },
    (p) => `follows=${p.followsCount || 0}, followers=${p.followersCount || 0}`,
  );

  if (!luna.did || !luna.accessJwt) {
    result.stepFailed("Search actors", "Luna account not created");
  } else {
    await timedCall(
      result,
      "Search actors",
      async () => {
        const res = await client.raw.get(
          "app.bsky.actor.searchActors",
          { q: "Luna" },
          luna.accessJwt,
        );
        return res;
      },
      (s) => `found=${s.actors?.length || 0}`,
    );
  }

  result.finish();
  return result;
}

if (import.meta.main) {
  run(createScenarioContext()).then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
